src/util/bip152.rs

```diff
@@ -0,0 +1,438 @@
+
+use hash_types::BlockHash;
+use blockdata::block::{Block, BlockHeader};
+use blockdata::transaction::Transaction;
+use consensus::encode::{self, Decodable, Encodable, VarInt};
+use util::endian;
+
+/// A BIP-152 error
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Error {
+    /// An unknown version number was used.
+    UnknownVersion,
+    /// A transaction index is requested that is out
+    /// of range from the corresponding block.
+    TxIndexOutOfRange(u64),
+    /// The prefill slice provided was invalid.
+    InvalidPrefill,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::UnknownVersion => write!(f, "an unknown version number was used"),
+            Error::TxIndexOutOfRange(i) => write!(f, "a transaction index is requested that is \
+                out of range from the corresponding block: {}", i,
+            ),
+            Error::InvalidPrefill => write!(f, "the prefill slice provided was invalid"),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn cause(&self) -> Option<&error::Error> {
+        None
+    }
+
+    fn description(&self) -> &str {
+        "description() is deprecated; use Display"
+    }
+}
+
+/// A [PrefilledTransaction] structure is used in [HeaderAndShortIDs] to
+/// provide a list of a few transactions explicitly.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct PrefilledTransaction {
+    /// The index of the transaction in the block.
+    ///
+    /// This field is differentially encoded relative to the previous
+    /// prefilled transaction as described as follows:
+    ///
+    /// > Several uses of CompactSize below are "differentially encoded". For
+    /// > these, instead of using raw indexes, the number encoded is the
+    /// > difference between the current index and the previous index, minus one.
+    /// > For example, a first index of 0 implies a real index of 0, a second
+    /// > index of 0 thereafter refers to a real index of 1, etc.
+    pub idx: u64,
+    /// The actual transaction.
+    pub tx: Transaction,
+}
+
+impl convert::AsRef<Transaction> for PrefilledTransaction {
+    fn as_ref(&self) -> &Transaction {
+        &self.tx
+    }
+}
+
+impl Encodable for PrefilledTransaction {
+    #[inline]
+    fn consensus_encode<S: io::Write>(&self, mut s: S) -> Result<usize, encode::Error> {
+        Ok(VarInt(self.idx).consensus_encode(&mut s)? + self.tx.consensus_encode(s)?)
+    }
+}
+
+impl Decodable for PrefilledTransaction {
+    #[inline]
+    fn consensus_decode<D: io::Read>(mut d: D) -> Result<PrefilledTransaction, encode::Error> {
+        let idx = VarInt::consensus_decode(&mut d)?.0;
+        let tx = Transaction::consensus_decode(d)?;
+        Ok(PrefilledTransaction { idx, tx })
+    }
+}
+
+/// Short transaction IDs are used to represent a transaction without sending a full 256-bit hash.
+#[derive(PartialEq, Eq, Clone, Default)]
```

Comment:
Might as well derive copy and derive hash so ShortId can be used as a key in a map (I need this)

```suggestion
#[derive(PartialEq, Eq, Clone, Default, Hash, Copy)]
```